metadatastore/commands.py

```diff
@@ -541,13 +550,137 @@ def _get_mongo_document(document, document_cls):
     mongo_document = document_cls.objects(__raw__={'uid': document})[0]
     return mongo_document
 
+def _dereference_uid_fields():
+    return
 
 @_ensure_connection
-def find_run_starts(**kwargs):
+def find_corrections(newest=False, dereference_uids=True, **kwargs):
+    """
+    Parameters
+    ----------
+    newest : bool, optional
+        True: only return the newest one
+    uid : str, optional
+        The uid of the original metadatastore Document, shared between all
+        instances of its Corrections
+    correction_uid : str, optional
+        The unique identifier for one Correction document in the Correction
+        collection
+    start_time : time-like, optional
+        time-like representation of the earliest time that a RunStart
+        was created. Valid options are:
+           - timestamps --> time.time()
+           - '2015'
+           - '2015-01'
+           - '2015-01-30'
+           - '2015-03-30 03:00:00'
+           - datetime.now()
+    stop_time : time-like, optional
+        timestamp of the latest time that a RunStart was created. See
+        docs for `start_time` for examples.
+    uid : str, optional
+        Globally unique id string provided to metadatastore
+    """
+    _normalize_object_id(kwargs, '_id')
+    _format_time(kwargs)
+    correction_dicts = _find_documents(Correction, **kwargs)
+    for correction_dict in correction_dicts:
+        if dereference_uids:
+            _dereference_reference_fields(correction_dict, newest=newest)
+        yield Document(correction_dict, correction_dict[
+            'original_document_type'] + ' -- Correction')
+
+
+def _dereference_reference_fields(mongo_dict, newest=True):
+    """In-place dereferencing
+    """
+    # reference_fields = {'run_start_id': find_run_starts,
+    #                     'run_start': find_run_starts,
+    #                     'descriptor_id': find_event_descriptors,
+    #                     'descriptor': find_event_descriptors,
+    #                     'beamline_config_id': find_beamline_configs,
+    #                     'beamline_config': find_beamline_configs}
+    reference_fields = {'run_start_id': RunStart,
+                        'run_start': RunStart,
+                        'descriptor_id': EventDescriptor,
+                        'descriptor': EventDescriptor,
+                        'beamline_config_id': BeamlineConfig,
+                        'beamline_config': BeamlineConfig}
+
+    for ref, klass in six.iteritems(reference_fields):
+        field = mongo_dict.get(ref, None)
+        if field:
+            # then field is either an ObjectId or a uid, or the object model
+            # is borked...
+            if isinstance(field, ObjectId):
+                # find the original document
+                kwargs = {'_id': field}
+            elif isinstance(field, six.text_type):
+                # it better be a uid
+                kwargs = {'uid': field}
+            else:
```

Comment:
Can you add a comment that this is to catch cases of already de-referenced documents.